### ISSUE TITLE
Limit the characters of the type text macro action input

### ISSUE THREAD

**Issue by mondalaci** [UltimateHackingKeyboard/agent#629]
Make it impossible to type anything other than letters, numbers, punctuation characters, and space characters (according to the content of the scancode select2 input) into as the text of the type text macro action.

**Comment by ert78gb** [comment]
Could you define the allowed characters as regular expression?
My plan to use the following regexp [A-Za-z0-9\s]|[[:punct:]] to check the keypressed charachter is valid or not.
If the user paste the content from clipboard the system will check the whole input text data and if it not valid by the regexp it not allow to paste the content.
My questions are:
- This solution is good for you
- should write any information to the user if paste not allowed?

**Comment by piratk** [comment]
I would also suggest to refrain from using a own regexp to define "valid" characters. What if I wanted a macro that prints "I åa ä e ö" (which is dialectal Swedish and means: "In the river there is an island")?

**Comment by mondalaci** [comment]
@eltang `[[;ascii:]]` may be the best solution by the name of it. Can't find the definition of these regexp character sets. Any help is welcome.

@piratk Dialetical marks will be thrown away, but you'll be able to invoke them using their non-US character equivalent in Agent. Keyboards send scancodes, not characters, so this is an extremely hard problem. I'll write a blog post about this instead of trying to explain it in depth here.

**Comment by eltang** [comment]
@mondalaci It's like [[:punct:]], but it's equivalent to the whole regex suggested by @ert78gb.

**Comment by ert78gb** [comment]
Just for the clarification. Is the whole ASCII character set supported ? Because the original issue contains a smaller set

**Comment by eltang** [comment]
@ert78gb Which characters do you have in mind? The only thing I can think of is \n, which should be supported.

**Comment by ert78gb** [comment]
I don't have exact character in my mind I just realised the
> letters, numbers, punctuation characters, and space characters

and the `[[:ascii:]]` allow every character between 0-127
https://www.asciitable.com/

**Comment by mondalaci** [comment]
@ert78gb Yes, please let's support every ASCII character. Sorry for not being very specific in the beginning.
